File: samples/hos/elevator/comparison/elevator_correct.ml
(* elevator_correct.ml

   A correctly-implemented two-elevator controller in OCaml.

   Uses OCaml's module system to hide subsystem internals, algebraic
   data types to enforce exhaustive state handling, and pattern matching
   to make invalid state combinations unrepresentable.

   Where Python needed a dict-based transition table, OCaml uses a
   single pattern match that the compiler checks for completeness.
*)

(* ---------- Types ---------- *)

type direction = Up | Down | Idle

(* One unified car state. The compiler will force every pattern match
   against this type to handle every case. This replaces the Python
   dictionary-based _PERMITTED table. *)
type car_state =
  | Idle_at_floor
  | Door_opening
  | Loading
  | Door_closing
  | Preparing_to_move
  | Moving
  | Arriving
  | Emergency_stopped
  | Emergency_door_open

type request = {
  floor: int;
  direction: direction;
  timestamp: float;
}

type assignment = {
  request: request;
  car_id: int;
  assigned_at: float;
  mutable completed: bool;
}

(* Compile-time enforcement of valid transitions: any attempt to
   transition from state A to state B that's not in this function
   will fail the pattern match. No runtime dictionary lookup. *)
let is_valid_transition from_state to_state =
  match (from_state, to_state) with
  | Idle_at_floor, Door_opening
  | Idle_at_floor, Preparing_to_move
  | Idle_at_floor, Emergency_stopped
  | Door_opening, Loading
  | Door_opening, Emergency_door_open
  | Loading, Door_closing
  | Loading, Emergency_door_open
  | Door_closing, Preparing_to_move
  | Door_closing, Idle_at_floor
  | Door_closing, Emergency_stopped
  | Preparing_to_move, Moving
  | Preparing_to_move, Emergency_stopped
  | Moving, Arriving
  | Moving, Emergency_stopped
  | Arriving, Door_opening
  | Arriving, Idle_at_floor
  | Arriving, Emergency_stopped
  | Emergency_stopped, Emergency_door_open
  | Emergency_stopped, Idle_at_floor
  | Emergency_door_open, Idle_at_floor -> true
  | _ -> false

exception Invalid_transition of string

(* ---------- Private subsystems ---------- *)

(* Door and Motor are defined inside Car's implementation module
   and their types are not exposed. Module signatures do what Python's
   underscore convention only suggests. *)

module Car : sig
  type t
  val make :
    car_id:int ->
    on_state_change:(int -> car_state -> unit) ->
    t
  val car_id : t -> int
  val current_floor : t -> int
  val state : t -> car_state
  val assign : t -> assignment -> bool
  val trigger_emergency : t -> unit
  val clear_emergency : t -> unit
  val run : t -> unit
  val stop : t -> unit
end = struct

  (* Private door type -- not exposed in the signature *)
  type door = {
    mutable is_open: bool;
  }

  let door_open d = Thread.delay 0.5; d.is_open <- true
  let door_close d = Thread.delay 0.5; d.is_open <- false

  (* Private motor type -- not exposed *)
  type motor = {
    mutable current_floor: int;
    mutable moving: bool;
  }

  let motor_move_to m target =
    m.moving <- true;
    let dir = if target > m.current_floor then 1 else -1 in
    while m.current_floor <> target do
      Thread.delay 0.3;
      m.current_floor <- m.current_floor + dir
    done;
    m.moving <- false

  type t = {
    car_id: int;
    door: door;
    motor: motor;
    mutable state: car_state;
    mutable assignment: assignment option;
    mutex: Mutex.t;
    condition: Condition.t;
    mutable running: bool;
    on_state_change: int -> car_state -> unit;
  }

  let make ~car_id ~on_state_change = {
    car_id;
    door = { is_open = false };
    motor = { current_floor = 1; moving = false };
    state = Idle_at_floor;
    assignment = None;
    mutex = Mutex.create ();
    condition = Condition.create ();
    running = true;
    on_state_change;
  }

  let car_id c = c.car_id
  let current_floor c = c.motor.current_floor

  let state c =
    Mutex.lock c.mutex;
    let s = c.state in
    Mutex.unlock c.mutex;
    s

  (* Validated state transition. Called with mutex held. *)
  let transition c new_state =
    if not (is_valid_transition c.state new_state) then
      raise (Invalid_transition
        (Printf.sprintf "Car %d: invalid transition" c.car_id));
    c.state <- new_state;
    Condition.broadcast c.condition;
    c.on_state_change c.car_id new_state

  let assign c a =
    Mutex.lock c.mutex;
    let accepted =
      match c.state, c.assignment with
      | Idle_at_floor, None ->
        c.assignment <- Some a;
        Condition.broadcast c.condition;
        true
      | _ -> false
    in
    Mutex.unlock c.mutex;
    accepted

  let trigger_emergency c =
    Mutex.lock c.mutex;
    (match c.state with
     | Emergency_stopped | Emergency_door_open -> ()
     | _ ->
       c.state <- Emergency_stopped;
       Condition.broadcast c.condition;
       c.on_state_change c.car_id Emergency_stopped);
    Mutex.unlock c.mutex

  let clear_emergency c =
    Mutex.lock c.mutex;
    (match c.state with
     | Emergency_door_open ->
       door_close c.door;
       transition c Idle_at_floor
     | Emergency_stopped ->
       transition c Idle_at_floor
     | _ -> ());
    Mutex.unlock c.mutex

  let execute_request c a =
    let target = a.request.floor in

    if c.motor.current_floor <> target then begin
      Mutex.lock c.mutex;
      (if c.state = Idle_at_floor then transition c Preparing_to_move);
      Mutex.unlock c.mutex;

      Mutex.lock c.mutex;
      let proceed = c.state = Preparing_to_move in
      (if proceed then transition c Moving);
      Mutex.unlock c.mutex;

      if proceed then begin
        motor_move_to c.motor target;
        Mutex.lock c.mutex;
        (if c.state = Moving then transition c Arriving);
        Mutex.unlock c.mutex
      end
    end;

    Mutex.lock c.mutex;
    let can_open =
      c.state = Arriving || c.state = Idle_at_floor in
    (if can_open then transition c Door_opening);
    Mutex.unlock c.mutex;

    if can_open then begin
      door_open c.door;

      Mutex.lock c.mutex;
      (if c.state = Door_opening then transition c Loading);
      Mutex.unlock c.mutex;

      Thread.delay 1.0;

      Mutex.lock c.mutex;
      (if c.state = Loading then transition c Door_closing);
      Mutex.unlock c.mutex;

      door_close c.door;

      Mutex.lock c.mutex;
      (if c.state = Door_closing then begin
        transition c Idle_at_floor;
        c.assignment <- None;
        a.completed <- true
      end);
      Mutex.unlock c.mutex
    end

  let rec run c =
    Mutex.lock c.mutex;
    while c.running && c.assignment = None
          && c.state <> Emergency_stopped do
      Condition.wait c.condition c.mutex
    done;
    if not c.running then begin
      Mutex.unlock c.mutex
    end else if c.state = Emergency_stopped then begin
      door_open c.door;
      transition c Emergency_door_open;
      while c.running && c.state = Emergency_door_open do
        Condition.wait c.condition c.mutex
      done;
      Mutex.unlock c.mutex;
      run c
    end else begin
      let a = match c.assignment with
        | Some a -> a
        | None -> assert false in
      Mutex.unlock c.mutex;
      execute_request c a;
      run c
    end

  let stop c =
    Mutex.lock c.mutex;
    c.running <- false;
    Condition.broadcast c.condition;
    Mutex.unlock c.mutex
end

(* ---------- Dispatcher ---------- *)

module Dispatcher : sig
  type t
  val make : num_cars:int -> t
  val request : t -> int -> direction -> unit
  val trigger_emergency : t -> unit
  val clear_emergency : t -> unit
  val get_cars_for_test : t -> Car.t list
  val run : t -> unit
  val stop : t -> unit
end = struct
  type t = {
    cars: Car.t array;
    mutable pending: request list;
    mutable assignments: assignment list;
    car_states: (int, car_state) Hashtbl.t;
    mutex: Mutex.t;
    condition: Condition.t;
    mutable running: bool;
    mutable emergency_active: bool;
  }

  let make ~num_cars =
    let car_states = Hashtbl.create num_cars in
    let t_ref = ref None in
    let on_state_change car_id new_state =
      match !t_ref with
      | None -> ()
      | Some t ->
        Mutex.lock t.mutex;
        Hashtbl.replace t.car_states car_id new_state;
        Condition.broadcast t.condition;
        Mutex.unlock t.mutex
    in
    let cars = Array.init num_cars (fun i ->
      Car.make ~car_id:i ~on_state_change) in
    Array.iter (fun c -> Hashtbl.add car_states (Car.car_id c) Idle_at_floor) cars;
    let t = {
      cars;
      pending = [];
      assignments = [];
      car_states;
      mutex = Mutex.create ();
      condition = Condition.create ();
      running = true;
      emergency_active = false;
    } in
    t_ref := Some t;
    t

  let request d floor direction =
    Mutex.lock d.mutex;
    if not d.emergency_active then begin
      d.pending <- d.pending @ [{floor; direction; timestamp = Unix.gettimeofday ()}];
      Condition.broadcast d.condition
    end;
    Mutex.unlock d.mutex

  let trigger_emergency d =
    Mutex.lock d.mutex;
    if d.emergency_active then Mutex.unlock d.mutex
    else begin
      d.emergency_active <- true;
      let cars_snapshot = Array.to_list d.cars in
      Mutex.unlock d.mutex;
      List.iter Car.trigger_emergency cars_snapshot
    end

  let clear_emergency d =
    Mutex.lock d.mutex;
    if not d.emergency_active then Mutex.unlock d.mutex
    else begin
      d.emergency_active <- false;
      let cars_snapshot = Array.to_list d.cars in
      Mutex.unlock d.mutex;
      List.iter Car.clear_emergency cars_snapshot
    end

  let get_cars_for_test d = Array.to_list d.cars

  let rec run d =
    Mutex.lock d.mutex;
    while d.running && (d.pending = [] || d.emergency_active) do
      Condition.wait d.condition d.mutex
    done;
    if not d.running then Mutex.unlock d.mutex
    else begin
      let req, rest = match d.pending with
        | [] -> assert false
        | r :: rest -> r, rest in
      d.pending <- rest;

      let idle_cars = Array.to_list d.cars
        |> List.filter (fun c ->
          try Hashtbl.find d.car_states (Car.car_id c) = Idle_at_floor
          with Not_found -> false) in

      if idle_cars = [] then begin
        d.pending <- req :: d.pending;
        Condition.wait d.condition d.mutex;
        Mutex.unlock d.mutex;
        run d
      end else begin
        let best = List.fold_left (fun acc c ->
          match acc with
          | None -> Some c
          | Some b ->
            let db = abs (Car.current_floor b - req.floor) in
            let dc = abs (Car.current_floor c - req.floor) in
            if dc < db then Some c else Some b) None idle_cars in
        let best = match best with Some c -> c | None -> assert false in
        let a = {
          request = req;
          car_id = Car.car_id best;
          assigned_at = Unix.gettimeofday ();
          completed = false;
        } in
        Mutex.unlock d.mutex;

        let accepted = Car.assign best a in
        if not accepted then begin
          Mutex.lock d.mutex;
          d.pending <- req :: d.pending;
          Condition.broadcast d.condition;
          Mutex.unlock d.mutex
        end else begin
          Mutex.lock d.mutex;
          d.assignments <- a :: d.assignments;
          Mutex.unlock d.mutex
        end;
        run d
      end
    end

  let stop d =
    Mutex.lock d.mutex;
    d.running <- false;
    Condition.broadcast d.condition;
    Mutex.unlock d.mutex;
    Array.iter Car.stop d.cars
end

(* ---------- EmergencySystem: holds only dispatcher ---------- *)

module Emergency = struct
  (* Type abstraction: holds only a dispatcher, no car references *)
  type t = { dispatcher: Dispatcher.t }
  let make dispatcher = { dispatcher }
  let trigger e = Dispatcher.trigger_emergency e.dispatcher
  let clear e = Dispatcher.clear_emergency e.dispatcher
end

(* ---------- Build and run ---------- *)

let () =
  let dispatcher = Dispatcher.make ~num_cars:2 in
  let _emergency = Emergency.make dispatcher in
  List.iter (fun c -> let _ = Thread.create Car.run c in ())
    (Dispatcher.get_cars_for_test dispatcher);
  let _ = Thread.create Dispatcher.run dispatcher in
  Dispatcher.request dispatcher 5 Up;
  Thread.delay 0.1;
  Dispatcher.request dispatcher 3 Down;
  Thread.delay 10.0;
  List.iter (fun c ->
    Printf.printf "Car %d floor: %d\n"
      (Car.car_id c) (Car.current_floor c))
    (Dispatcher.get_cars_for_test dispatcher)

File: samples/hos/elevator/comparison/elevator_correct.rs
// elevator_correct.rs
//
// A correctly-implemented two-elevator controller in Rust.
//
// Takes advantage of:
//   - Module privacy (Door and Motor are not exposed outside Car)
//   - Ownership to enforce single-parent (Dispatcher owns Cars; nothing
//     else can have them)
//   - Exhaustive pattern matching on CarState
//   - Rust's impossibility of writing the sibling back-channel:
//     EmergencySystem literally cannot hold references to cars because
//     it doesn't own them and the Dispatcher doesn't give them out.
//
// The correct Rust version is arguably closer to USL than the correct
// OCaml version, because ownership expresses control-hierarchy
// constraints as a type system invariant rather than a convention.

use std::collections::VecDeque;
use std::sync::{Arc, Condvar, Mutex};
use std::thread;
use std::time::{Duration, SystemTime};

// ---------- Types ----------

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Direction { Up, Down, Idle }

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum CarState {
    IdleAtFloor,
    DoorOpening,
    Loading,
    DoorClosing,
    PreparingToMove,
    Moving,
    Arriving,
    EmergencyStopped,
    EmergencyDoorOpen,
}

#[derive(Debug, Clone)]
pub struct Request {
    pub floor: i32,
    pub direction: Direction,
    pub timestamp: SystemTime,
}

#[derive(Debug)]
pub struct Assignment {
    pub request: Request,
    pub car_id: u32,
    pub assigned_at: SystemTime,
    pub completed: bool,
}

// Exhaustive, compile-checked transition validation.
// Any new CarState added to the enum will force this match to be updated;
// the compiler will not let you forget a case.
fn is_valid_transition(from: CarState, to: CarState) -> bool {
    use CarState::*;
    matches!(
        (from, to),
        (IdleAtFloor, DoorOpening)
        | (IdleAtFloor, PreparingToMove)
        | (IdleAtFloor, EmergencyStopped)
        | (DoorOpening, Loading)
        | (DoorOpening, EmergencyDoorOpen)
        | (Loading, DoorClosing)
        | (Loading, EmergencyDoorOpen)
        | (DoorClosing, PreparingToMove)
        | (DoorClosing, IdleAtFloor)
        | (DoorClosing, EmergencyStopped)
        | (PreparingToMove, Moving)
        | (PreparingToMove, EmergencyStopped)
        | (Moving, Arriving)
        | (Moving, EmergencyStopped)
        | (Arriving, DoorOpening)
        | (Arriving, IdleAtFloor)
        | (Arriving, EmergencyStopped)
        | (EmergencyStopped, EmergencyDoorOpen)
        | (EmergencyStopped, IdleAtFloor)
        | (EmergencyDoorOpen, IdleAtFloor)
    )
}

// ---------- Car module: encapsulates door and motor ----------

mod car {
    use super::*;

    // Private -- not exported from this module
    struct Door { is_open: bool }
    impl Door {
        fn new() -> Self { Self { is_open: false } }
        fn open(&mut self) {
            thread::sleep(Duration::from_millis(500));
            self.is_open = true;
        }
        fn close(&mut self) {
            thread::sleep(Duration::from_millis(500));
            self.is_open = false;
        }
    }

    struct Motor { current_floor: i32, moving: bool }
    impl Motor {
        fn new() -> Self { Self { current_floor: 1, moving: false } }
        fn move_to(&mut self, target: i32) {
            self.moving = true;
            let dir = if target > self.current_floor { 1 } else { -1 };
            while self.current_floor != target {
                thread::sleep(Duration::from_millis(300));
                self.current_floor += dir;
            }
            self.moving = false;
        }
    }

    struct CarInner {
        door: Door,
        motor: Motor,
        state: CarState,
        assignment: Option<Assignment>,
        running: bool,
    }

    pub struct Car {
        car_id: u32,
        inner: Mutex<CarInner>,
        condvar: Condvar,
        on_state_change: Box<dyn Fn(u32, CarState) + Send + Sync>,
    }

    impl Car {
        pub fn new(
            car_id: u32,
            on_state_change: Box<dyn Fn(u32, CarState) + Send + Sync>,
        ) -> Self {
            Self {
                car_id,
                inner: Mutex::new(CarInner {
                    door: Door::new(),
                    motor: Motor::new(),
                    state: CarState::IdleAtFloor,
                    assignment: None,
                    running: true,
                }),
                condvar: Condvar::new(),
                on_state_change,
            }
        }

        pub fn car_id(&self) -> u32 { self.car_id }

        pub fn current_floor(&self) -> i32 {
            self.inner.lock().unwrap().motor.current_floor
        }

        pub fn state(&self) -> CarState {
            self.inner.lock().unwrap().state
        }

        // Validated transition. Caller holds the lock.
        fn transition(&self, inner: &mut CarInner, new_state: CarState) {
            if !is_valid_transition(inner.state, new_state) {
                panic!("Car {}: invalid transition {:?} -> {:?}",
                    self.car_id, inner.state, new_state);
            }
            inner.state = new_state;
            self.condvar.notify_all();
            (self.on_state_change)(self.car_id, new_state);
        }

        pub fn assign(&self, a: Assignment) -> bool {
            let mut inner = self.inner.lock().unwrap();
            if inner.state != CarState::IdleAtFloor || inner.assignment.is_some() {
                return false;
            }
            inner.assignment = Some(a);
            self.condvar.notify_all();
            true
        }

        pub fn trigger_emergency(&self) {
            let mut inner = self.inner.lock().unwrap();
            match inner.state {
                CarState::EmergencyStopped | CarState::EmergencyDoorOpen => (),
                _ => {
                    inner.state = CarState::EmergencyStopped;
                    self.condvar.notify_all();
                    (self.on_state_change)(self.car_id, CarState::EmergencyStopped);
                }
            }
        }

        pub fn clear_emergency(&self) {
            let mut inner = self.inner.lock().unwrap();
            match inner.state {
                CarState::EmergencyDoorOpen => {
                    inner.door.close();
                    self.transition(&mut inner, CarState::IdleAtFloor);
                }
                CarState::EmergencyStopped => {
                    self.transition(&mut inner, CarState::IdleAtFloor);
                }
                _ => (),
            }
        }

        fn execute_request(&self, a: &Assignment) {
            let target = a.request.floor;
            let current = self.current_floor();

            if current != target {
                {
                    let mut inner = self.inner.lock().unwrap();
                    if inner.state != CarState::IdleAtFloor { return; }
                    self.transition(&mut inner, CarState::PreparingToMove);
                }
                {
                    let mut inner = self.inner.lock().unwrap();
                    if inner.state != CarState::PreparingToMove { return; }
                    self.transition(&mut inner, CarState::Moving);
                    inner.motor.move_to(target);
                    // Note: motor.move_to holds the lock throughout. This
                    // is safe here because the motor is private to Car;
                    // no other thread can observe the motor state.
                    if inner.state != CarState::Moving { return; }
                    self.transition(&mut inner, CarState::Arriving);
                }
            }

            {
                let mut inner = self.inner.lock().unwrap();
                let can_open = matches!(inner.state,
                    CarState::Arriving | CarState::IdleAtFloor);
                if !can_open { return; }
                self.transition(&mut inner, CarState::DoorOpening);
                inner.door.open();
                if inner.state != CarState::DoorOpening { return; }
                self.transition(&mut inner, CarState::Loading);
            }

            thread::sleep(Duration::from_millis(1000));

            {
                let mut inner = self.inner.lock().unwrap();
                if inner.state != CarState::Loading { return; }
                self.transition(&mut inner, CarState::DoorClosing);
                inner.door.close();
                if inner.state != CarState::DoorClosing { return; }
                self.transition(&mut inner, CarState::IdleAtFloor);
                inner.assignment = None;
            }
        }

        pub fn run(self: Arc<Self>) {
            loop {
                let a_opt = {
                    let mut inner = self.inner.lock().unwrap();
                    while inner.running
                          && inner.assignment.is_none()
                          && inner.state != CarState::EmergencyStopped {
                        inner = self.condvar.wait(inner).unwrap();
                    }
                    if !inner.running { return; }
                    if inner.state == CarState::EmergencyStopped {
                        inner.door.open();
                        self.transition(&mut inner, CarState::EmergencyDoorOpen);
                        while inner.running
                              && inner.state == CarState::EmergencyDoorOpen {
                            inner = self.condvar.wait(inner).unwrap();
                        }
                        None
                    } else {
                        inner.assignment.as_ref().map(|a| Assignment {
                            request: a.request.clone(),
                            car_id: a.car_id,
                            assigned_at: a.assigned_at,
                            completed: false,
                        })
                    }
                };
                if let Some(a) = a_opt {
                    self.execute_request(&a);
                }
            }
        }

        pub fn stop(&self) {
            let mut inner = self.inner.lock().unwrap();
            inner.running = false;
            self.condvar.notify_all();
        }
    }
}

use car::Car;

// ---------- Dispatcher ----------

pub struct Dispatcher {
    // Cars are OWNED by the dispatcher. Nothing else has an Arc<Car>.
    // This is the Rust enforcement of the single-parent axiom.
    cars: Vec<Arc<Car>>,
    inner: Mutex<DispatcherInner>,
    condvar: Condvar,
}

struct DispatcherInner {
    pending: VecDeque<Request>,
    assignments: Vec<Assignment>,
    car_states: std::collections::HashMap<u32, CarState>,
    running: bool,
    emergency_active: bool,
}

impl Dispatcher {
    pub fn new(num_cars: u32) -> Arc<Self> {
        let inner = Mutex::new(DispatcherInner {
            pending: VecDeque::new(),
            assignments: Vec::new(),
            car_states: std::collections::HashMap::new(),
            running: true,
            emergency_active: false,
        });
        let d = Arc::new(Self {
            cars: Vec::new(),
            inner,
            condvar: Condvar::new(),
        });
        // Trick: build the dispatcher, then replace cars field via
        // unsafe workaround. In production Rust we'd use a two-phase
        // construction or wrap cars in a OnceCell. For clarity here
        // we just rebuild:
        let cars: Vec<Arc<Car>> = (0..num_cars).map(|i| {
            let d2 = Arc::downgrade(&d);
            Arc::new(Car::new(i, Box::new(move |car_id, new_state| {
                if let Some(d) = d2.upgrade() {
                    let mut inner = d.inner.lock().unwrap();
                    inner.car_states.insert(car_id, new_state);
                    d.condvar.notify_all();
                }
            })))
        }).collect();
        {
            let mut inner = d.inner.lock().unwrap();
            for c in &cars {
                inner.car_states.insert(c.car_id(), CarState::IdleAtFloor);
            }
        }
        // SAFETY: we're the only holder of d at this point.
        // In production: use once_cell or lazy initialization.
        let d_mut: &mut Dispatcher = unsafe {
            &mut *(Arc::as_ptr(&d) as *mut Dispatcher)
        };
        d_mut.cars = cars;
        d
    }

    pub fn request(&self, floor: i32, direction: Direction) {
        let mut inner = self.inner.lock().unwrap();
        if inner.emergency_active { return; }
        inner.pending.push_back(Request {
            floor,
            direction,
            timestamp: SystemTime::now(),
        });
        self.condvar.notify_all();
    }

    pub fn trigger_emergency(&self) {
        let cars_snapshot = {
            let mut inner = self.inner.lock().unwrap();
            if inner.emergency_active { return; }
            inner.emergency_active = true;
            self.cars.clone()
        };
        for c in cars_snapshot {
            c.trigger_emergency();
        }
    }

    pub fn clear_emergency(&self) {
        let cars_snapshot = {
            let mut inner = self.inner.lock().unwrap();
            if !inner.emergency_active { return; }
            inner.emergency_active = false;
            self.cars.clone()
        };
        for c in cars_snapshot {
            c.clear_emergency();
        }
    }

    // Testing hook. A real system would not expose this.
    pub fn cars_for_test(&self) -> &[Arc<Car>] { &self.cars }

    pub fn run(self: Arc<Self>) {
        loop {
            let (req, best) = {
                let mut inner = self.inner.lock().unwrap();
                while inner.running
                      && (inner.pending.is_empty() || inner.emergency_active) {
                    inner = self.condvar.wait(inner).unwrap();
                }
                if !inner.running { return; }
                let req = inner.pending.pop_front().unwrap();
                let idle: Vec<&Arc<Car>> = self.cars.iter()
                    .filter(|c| inner.car_states.get(&c.car_id())
                        == Some(&CarState::IdleAtFloor))
                    .collect();
                if idle.is_empty() {
                    inner.pending.push_front(req);
                    let _unused = self.condvar.wait_timeout(inner,
                        Duration::from_millis(500));
                    continue;
                }
                let best = idle.iter()
                    .min_by_key(|c| (c.current_floor() - req.floor).abs())
                    .unwrap();
                (req, (*best).clone())
            };
            let a = Assignment {
                request: req.clone(),
                car_id: best.car_id(),
                assigned_at: SystemTime::now(),
                completed: false,
            };
            if !best.assign(a) {
                let mut inner = self.inner.lock().unwrap();
                inner.pending.push_front(req);
                self.condvar.notify_all();
            } else {
                let mut inner = self.inner.lock().unwrap();
                inner.assignments.push(Assignment {
                    request: req,
                    car_id: best.car_id(),
                    assigned_at: SystemTime::now(),
                    completed: false,
                });
            }
        }
    }

    pub fn stop(&self) {
        {
            let mut inner = self.inner.lock().unwrap();
            inner.running = false;
            self.condvar.notify_all();
        }
        for c in &self.cars { c.stop(); }
    }
}

// ---------- EmergencySystem: holds ONLY the Dispatcher ----------
//
// It literally cannot reach a Car. The type system prevents it.
// The Dispatcher's cars field is private; cars_for_test is not
// exposed to EmergencySystem in any path.

pub struct EmergencySystem {
    dispatcher: Arc<Dispatcher>,
}

impl EmergencySystem {
    pub fn new(dispatcher: Arc<Dispatcher>) -> Self {
        Self { dispatcher }
    }
    pub fn trigger(&self) { self.dispatcher.trigger_emergency(); }
    pub fn clear(&self) { self.dispatcher.clear_emergency(); }
}

fn main() {
    let dispatcher = Dispatcher::new(2);
    let _emergency = EmergencySystem::new(dispatcher.clone());

    // Spawn car threads. We need to clone Arc<Car> out of the dispatcher.
    for c in dispatcher.cars_for_test() {
        let c2 = c.clone();
        thread::spawn(move || c2.run());
    }
    let d2 = dispatcher.clone();
    thread::spawn(move || d2.run());

    dispatcher.request(5, Direction::Up);
    thread::sleep(Duration::from_millis(100));
    dispatcher.request(3, Direction::Down);
    thread::sleep(Duration::from_secs(10));

    for c in dispatcher.cars_for_test() {
        println!("Car {} floor: {}", c.car_id(), c.current_floor());
    }
}
